Add options validation source generation conceptual doc (#37681)
dotnet/docs: docs/core/extensions/options-validation-generator.md

---
title: Compile-time options validation source generation
author: tarekgh
description: Learn how to utilize compile-time source generation for validating options.
ms.author: tarekms
ms.date: 10/23/2023
---

# Compile-time options validation source generation

> In the [options pattern](options.md), various methods for validating options are presented. These methods include using data annotation attributes or employing a custom validator. Data annotation attributes are validated at runtime and can incur performance costs. This article demonstrates how to utilize the options validation source generator to produce optimized validation code during compile time.

Reviewer: ```suggestion
In the [options pattern](options.md), various methods for validating options are presented. These methods include using data annotation attributes or employing a custom validator. Data annotation attributes are validated at run time and can incur performance costs. This article demonstrates how to utilize the options validation source generator to produce optimized validation code at compile time.
```

## Automatic IValidateOptions implementation generation

> In the [options pattern](options.md), we've illustrated how to implement the IValidateOptions interface for validating options. The options validation source generator can automatically create the IValidateOptions interface implementation by leveraging data annotation attributes on the options class.

Reviewer: ```suggestion
The [options pattern](options.md) article illustrates how to implement the `IValidateOptions` interface for validating options. The options validation source generator can automatically create the `IValidateOptions` interface implementation by leveraging data annotation attributes on the options class.
```

> Let's take previous annotation attributes example which mentioned in [options pattern](options.md) to convert it to use the options validation source generator.

Reviewer: ```suggestion
The content that follows takes the annotation attributes example that's shown in [Options pattern](options.md) and converts it to use the options validation source generator.
```

Reviewer: Style guide says to avoid "us" and "we".

Consider the following *appsettings.json* file:

```json
{
  "MyCustomSettingsSection": {
    "SiteTitle": "Amazing docs from Awesome people!",
    "Scale": 10,
    "VerbosityLevel": 32
  }
}
```

The following class binds to the `"MyCustomSettingsSection"` configuration section and applies a couple of `DataAnnotations` rules:

:::code language="csharp" source="snippets/configuration/console-json/SettingsOptions.cs":::

> In the preceding `SettingsOptions` class, the `ConfigurationSectionName` property contains the name of the configuration section to bind to. In this scenario, the options object provides the name of its configuration section. Then we call the following code to bind the configuration section to the options object and validate the data annotations:

Reviewer: ```suggestion
In the preceding `SettingsOptions` class, the `ConfigurationSectionName` property contains the name of the configuration section to bind to. In this scenario, the options object provides the name of its configuration section. Then, call the following code to bind the configuration section to the options object and validate the data annotations:
```

```csharp
builder.Services
    .AddOptions<SettingsOptions>()
    .Bind(Configuration.GetSection(SettingsOptions.ConfigurationSectionName))
    .ValidateDataAnnotations();
```

> While this approach works effectively, it can introduce performance implications because validation occurs at runtime via reflection. Additionally, reliance on reflection can lead to issues, particularly when using AOT compilation. Leveraging compile-time source generation for options validation allows you to generate performance-optimized validation code and eliminates the need for reflection, resulting in smoother AOT-compatible app building. The following code demonstrates how to use the options validation source generator:

Reviewer: ```suggestion
While this approach is effective, it can affect performance because validation occurs at run time via reflection. Additionally, reliance on reflection can lead to issues, particularly when using AOT compilation. By leveraging compile-time source generation for options validation, you can generate performance-optimized validation code and eliminate the need for reflection, resulting in smoother AOT-compatible app building. The following code demonstrates how to use the options validation source generator:
```

```csharp
[OptionsValidator]
public partial class ValidateSettingsOptions : IValidateOptions<SettingsOptions>
{
}
```

> Employing <xref:Microsoft.Extensions.Options.OptionsValidatorAttribute> on an empty partial class instructs the options validation source generator to create the IValidateOptions interface implementation which validate `SettingsOptions`. The generated code by the options validation source generator will resemble the following example:

Reviewer: ```suggestion
The presence of the <xref:Microsoft.Extensions.Options.OptionsValidatorAttribute> on an empty partial class instructs the options validation source generator to create the `IValidateOptions` interface implementation that validates `SettingsOptions`. The code that's generated by the options validation source generator will resemble the following example:
```

                (MaximumIsExclusive ? max.CompareTo(convertedValue) > 0 : max.CompareTo(convertedValue) >= 0);
        }
        private string GetValidationErrorMessage()
        {
            return (MinimumIsExclusive, MaximumIsExclusive) switch
            {
                (false, false) => "The field {0} must be between {1} and {2}.",
                (true, false) => "The field {0} must be between {1} exclusive and {2}.",
                (false, true) => "The field {0} must be between {1} and {2} exclusive.",
                (true, true) => "The field {0} must be between {1} exclusive and {2} exclusive.",
            };
        }
        private object? ConvertValue(object? value, System.Globalization.CultureInfo formatProvider)
        {
            if (value is string stringValue)
            {
                value = global::System.Convert.ChangeType(stringValue, OperandType, formatProvider);
            }
            else
            {
                value = global::System.Convert.ChangeType(value, OperandType, formatProvider);
            }
            return value;
        }
    }
}

```

> The generated code is optimized for performance and doesn't rely on reflection. The generated code is also AOT-compatible. The generated code is placed in a file named *Validators.g.cs*.

Reviewer: ```suggestion
The generated code is optimized for performance and doesn't rely on reflection. It's also AOT-compatible. The generated code is placed in a file named *Validators.g.cs*.
```

> [!NOTE]

> > Users don't need to take any additional steps to enable the options validation source generator. It is automatically enabled by default when referencing [Microsoft.Extensions.Options](https://www.nuget.org/packages/Microsoft.Extensions.Options) version 8 and above in the project or when building an ASP.NET application.

Reviewer: ```suggestion
> You don't need to take any additional steps to enable the options validation source generator. It's automatically enabled by default when your project references [Microsoft.Extensions.Options](https://www.nuget.org/packages/Microsoft.Extensions.Options) version 8 or later, or when building an ASP.NET application.
```

> The only step the app developer needs to take is to add the following to the startup code:

Reviewer: ```suggestion
The only step you need to take is to add the following to the startup code:
```

```csharp
builder.Services
    .AddSingleton<IValidateOptions<SettingsOptions>, ValidateSettingsOptions>();
```

> [!NOTE]
> Calling `ValidateDataAnnotations` is not required when using the options validation source generator.

> When the application attempts to access the options object, the generated code for options validation will be executed to validate the options object. The following code snippet illustrates how to access the options object:

Reviewer: ```suggestion
When the application attempts to access the options object, the generated code for options validation is executed to validate the options object. The following code snippet illustrates how to access the options object:
```

```csharp

>     var settingsOptions = app.Services.GetRequiredService<IOptions<SettingsOptions>>().Value;

Reviewer: ```suggestion
var settingsOptions = app.Services.GetRequiredService<IOptions<SettingsOptions>>().Value;
```

```

## Replaced data annotation attributes

> Upon close examination of the generated code, you'll observe that the original data annotation attributes, such as <xref:System.ComponentModel.DataAnnotations.RangeAttribute> which were initially applied to the property `SettingsOptions.Scale`, have been substituted with custom attributes like `__SourceGen__RangeAttribute`. This substitution is made because the `RangeAttribute` relies on reflection for validation. In contrast, `__SourceGen__RangeAttribute` is a custom attribute optimized for performance and doesn't depend on reflection, making the code AOT-compatible. The same pattern of attribute replacement will be applied on <xref:System.ComponentModel.DataAnnotations.MaxLengthAttribute>, <xref:System.ComponentModel.DataAnnotations.MinLengthAttribute>, and <xref:System.ComponentModel.DataAnnotations.LengthAttribute> in addition to <xref:System.ComponentModel.DataAnnotations.RangeAttribute>.

Reviewer: ```suggestion
Upon close examination of the generated code, you'll observe that the original data annotation attributes, such as <xref:System.ComponentModel.DataAnnotations.RangeAttribute>, that were initially applied to the property `SettingsOptions.Scale`, have been substituted with custom attributes like `__SourceGen__RangeAttribute`. This substitution is made because the `RangeAttribute` relies on reflection for validation. In contrast, `__SourceGen__RangeAttribute` is a custom attribute optimized for performance and doesn't depend on reflection, making the code AOT-compatible. The same pattern of attribute replacement will be applied on <xref:System.ComponentModel.DataAnnotations.MaxLengthAttribute>, <xref:System.ComponentModel.DataAnnotations.MinLengthAttribute>, and <xref:System.ComponentModel.DataAnnotations.LengthAttribute> in addition to <xref:System.ComponentModel.DataAnnotations.RangeAttribute>.
```

> For anyone developing custom data annotation attributes, it is advisable to refrain from using reflection for validation. Instead, it's recommended to craft strongly typed code that does not rely on reflection. This approach ensures smooth compatibility with AOT builds.

Reviewer: ```suggestion
For anyone developing custom data annotation attributes, it's advisable to refrain from using reflection for validation. Instead, it's recommended to craft strongly typed code that doesn't rely on reflection. This approach ensures smooth compatibility with AOT builds.
```
